<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Standard analysis of ChIP-seq experiments — opening_act2 • hwglabr2</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://maxcdn.bootstrapcdn.com/bootswatch/3.3.7/yeti/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">

<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">


<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script>
<script src="../pkgdown.js"></script>

<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">hwglabr2</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../reference/index.html">Functions</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/hochwagenlab/hwglabr2">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

      <div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Standard analysis of ChIP-seq experiments</h1>
    </div>

    
    <p>Runs the lab's standard analysis of ChIP-seq experiment data and produces
several .pdf files of analysis plots to be kept in a new folder at
<code>.../LabShare/HTGenomics/Opening_act/</code>.</p>
    

    <pre class="usage"><span class='fu'>opening_act2</span>(<span class='no'>signal_data</span>, <span class='no'>genome</span>, <span class='no'>genotype</span>, <span class='no'>chip_target</span>, <span class='no'>sample_id</span>,
  <span class='kw'>output_path</span> <span class='kw'>=</span> <span class='st'>"/Volumes/LabShare/HTGenomics/Opening_act/"</span>,
  <span class='kw'>user_input</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>run_chr_size_bias</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>run_centromeres</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>run_rDNA</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>run_telomeres</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>run_dsb_hotspots</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>run_axis</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>run_meta_orf</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a> Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>signal_data</th>
      <td><p>Input signal track data as a <code>GRanges</code> object (see
<code>?"GRanges-class"</code> for more details). To load wiggle and bedGraph data
run <code><a href='import_wiggle.html'>import_wiggle</a></code> and <code><a href='import_bedGraph.html'>import_bedGraph</a></code>,
respectively. No default.</p></td>
    </tr>
    <tr>
      <th>genome</th>
      <td><p>Character object specifying the genome version; accepts one of
the following options:</p><ol>
<li><p><code>"SK1Yue"</code></p></li>
<li><p><code>"sacCer3"</code></p></li>
</ol><p>No default.</p></td>
    </tr>
    <tr>
      <th>genotype</th>
      <td><p>Character object indicating the relevant strain mutations.
Just use <code>genotype=WT</code>, for example, if there are no relevant mutations.
No default.</p></td>
    </tr>
    <tr>
      <th>chip_target</th>
      <td><p>Character object indicating the ChIP target protein.
No default.</p></td>
    </tr>
    <tr>
      <th>sample_id</th>
      <td><p>Character object indicating the sample ID, including the ID
used in the analysis pipeline (with a date) and the read mapping conditions
(see examples below). If <code>user_input=TRUE</code> the function asks the user to
check that the provided <code>sample_id</code> matches the required format before
proceeding with the analysis. No default.</p></td>
    </tr>
    <tr>
      <th>output_path</th>
      <td><p>Character object with a valid path to directory to save
output files at.
Defaults to <code>output_path='/Volumes/LabShare/HTGenomics/Opening_act/'</code></p></td>
    </tr>
    <tr>
      <th>user_input</th>
      <td><p>Logical indicating whether to ask user to check the format
of the <code>sample_id</code> argument. Defaults to <code>TRUE</code>.</p></td>
    </tr>
    <tr>
      <th>run_chr_size_bias</th>
      <td><p>Logical indicating whether to run the chromosome
size bias analysis. All included analysis can be turned off in this way.
Defaults to <code>TRUE</code>.</p></td>
    </tr>
    <tr>
      <th>run_centromeres</th>
      <td><p>Logical indicating whether to run the signal at 
centromere analysis. Defaults to <code>TRUE</code>.</p></td>
    </tr>
    <tr>
      <th>run_rDNA</th>
      <td><p>Logical indicating whether to run the signal flanking rDNA
analysis. Defaults to <code>TRUE</code>.</p></td>
    </tr>
    <tr>
      <th>run_telomeres</th>
      <td><p>Logical indicating whether to run the signal at
sub-telomeric region analysis. Defaults to <code>TRUE</code>.</p></td>
    </tr>
    <tr>
      <th>run_dsb_hotspots</th>
      <td><p>Logical indicating whether to run the signal at DSB
hotspot analysis. Defaults to <code>TRUE</code>.</p></td>
    </tr>
    <tr>
      <th>run_axis</th>
      <td><p>Logical indicating whether to run the signal at axis binding
site analysis. Defaults to <code>TRUE</code>.</p></td>
    </tr>
    <tr>
      <th>run_meta_orf</th>
      <td><p>Logical indicating whether to run the meta ORF analysis.
Defaults to <code>TRUE</code>.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A new folder at the specified location (default is
<code>.../LabShare/HTGenomics/Opening_act/</code>) containing output plots (as .pdf
files) of the included of the following analyses:</p><ol>
<li><p>Chromosome size bias</p></li>
<li><p>Signal at centromeres</p></li>
<li><p>Signal flanking rDNA</p></li>
<li><p>Signal at sub-telomeric regions</p></li>
<li><p>Signal at DSB hotspots</p></li>
<li><p>Signal at axis binding sites</p></li>
<li><p>Signal at meta ORF</p></li>
</ol>

    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='fu'>not_run</span>({
  <span class='fu'>opening_act2</span>(<span class='kw'>signal_data</span><span class='kw'>=</span><span class='no'>WT</span>, <span class='kw'>genome</span><span class='kw'>=</span><span class='st'>"sacCer3"</span>, <span class='kw'>genotype</span><span class='kw'>=</span><span class='st'>"WT"</span>,
               <span class='kw'>chip_target</span><span class='kw'>=</span><span class='st'>"Red1"</span>, <span class='kw'>sample_id</span><span class='kw'>=</span><span class='st'>"AH119C-040114-sacCer3-2mis"</span>)

  <span class='fu'>opening_act2</span>(<span class='no'>set1_wiggle</span>, <span class='kw'>genome</span><span class='kw'>=</span><span class='st'>"sacCer3"</span>, <span class='st'>"set1"</span>, <span class='st'>"Red1"</span>,
               <span class='st'>"AH8584b-16032016-sacCer3-2mis"</span>)

  <span class='fu'>opening_act2</span>(<span class='no'>rec8</span>, <span class='kw'>genome</span><span class='kw'>=</span><span class='st'>"SK1Yue"</span>, <span class='st'>"rec8"</span>, <span class='st'>"Red1"</span>,
               <span class='st'>"AH8115b-24042015-SK1Yue-PM"</span>, <span class='kw'>user_input</span><span class='kw'>=</span><span class='fl'>FALSE</span>,
               <span class='kw'>run_meta_orf</span><span class='kw'>=</span><span class='fl'>FALSE</span>)

  <span class='fu'>opening_act2</span>(<span class='kw'>signal_data</span><span class='kw'>=</span><span class='no'>WT</span>, <span class='kw'>genome</span><span class='kw'>=</span><span class='st'>"sacCer3"</span>, <span class='kw'>genotype</span><span class='kw'>=</span><span class='st'>"WT"</span>,
               <span class='kw'>chip_target</span><span class='kw'>=</span><span class='st'>"Red1"</span>, <span class='kw'>sample_id</span><span class='kw'>=</span><span class='st'>"AH119C-040114-sacCer3-2mis"</span>,
               <span class='kw'>run_meta_orf</span><span class='kw'>=</span><span class='fl'>FALSE</span>)

  <span class='fu'>opening_act2</span>(<span class='kw'>signal_data</span><span class='kw'>=</span><span class='no'>WT</span>, <span class='kw'>genome</span><span class='kw'>=</span><span class='st'>"sacCer3"</span>, <span class='kw'>genotype</span><span class='kw'>=</span><span class='st'>"dot1"</span>,
               <span class='kw'>chip_target</span><span class='kw'>=</span><span class='st'>"Red1"</span>, <span class='kw'>sample_id</span><span class='kw'>=</span><span class='st'>"AH8104-010116-sacCer3-2mis"</span>,
               <span class='kw'>output_path</span><span class='kw'>=</span><span class='st'>'~/Desktop'</span>,
               <span class='kw'>run_chr_size_bias</span><span class='kw'>=</span><span class='fl'>FALSE</span>, <span class='kw'>run_centromeres</span><span class='kw'>=</span><span class='fl'>FALSE</span>, <span class='kw'>run_rDNA</span><span class='kw'>=</span><span class='fl'>FALSE</span>,
               <span class='kw'>run_telomeres</span><span class='kw'>=</span><span class='fl'>FALSE</span>, <span class='kw'>run_dsb_hotspots</span><span class='kw'>=</span><span class='fl'>TRUE</span>, <span class='kw'>run_axis</span><span class='kw'>=</span><span class='fl'>TRUE</span>,
               <span class='kw'>run_meta_orf</span><span class='kw'>=</span><span class='fl'>FALSE</span>)
})</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Luis Vale Silva.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  </body>
</html>
