<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Signal around provided genomic positions — signal_at_position • hwglabr2</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">


<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script>
<script src="../pkgdown.js"></script>

<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">hwglabr2</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/hochwagenlab/hwglabr2">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

      <div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Signal around provided genomic positions</h1>
    </div>

    
    <p>Pulls out signal in intervals of the defined length centered at provided
positions.

    <strong>Note:</strong> Our ChIP-seq data always contains gaps with missing data. The
affected intervals will contain "NA" values in the output. As a warning, the
number of affected genes is printed to the console.</p>
    

    <pre class="usage"><span class='fu'>signal_at_position</span>(<span class='no'>signal_data</span>, <span class='no'>positions</span>, <span class='no'>position_names</span>, <span class='kw'>up_ext</span> <span class='kw'>=</span> <span class='fl'>1000</span>,
  <span class='kw'>down_ext</span> <span class='kw'>=</span> <span class='fl'>1000</span>, <span class='kw'>write_to_file</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a> Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>signal_data</th>
      <td><p>Input signal track data as a <code>GRanges</code> object (see
<code>?"GRanges-class"</code> for more details). To load wiggle and bedGraph data
run <code><a href='import_wiggle.html'>import_wiggle</a></code> and <code><a href='import_bedGraph.html'>import_bedGraph</a></code>,
respectively. No default.</p></td>
    </tr>
    <tr>
      <th>positions</th>
      <td><p>Position data as a <code>GRanges</code> object. Will also accept
an interval <code>GRanges</code> object of constant width (extended by specified
<code>up_ext</code> and <code>down_ext</code>). No default.</p></td>
    </tr>
    <tr>
      <th>position_names</th>
      <td><p>String indicating the column containing the position
names: either the <code>GRanges</code> 'seqnames' or a metadata column. Will be
added to the output to identify the different positions. No default.</p></td>
    </tr>
    <tr>
      <th>up_ext</th>
      <td><p>Integer indicating number of bp to collect upstream of start
position. Defaults to <code>1000</code>.</p></td>
    </tr>
    <tr>
      <th>down_ext</th>
      <td><p>Integer indicating number of bp to collect downstream of end
position. Deafaults to <code>1000</code>.</p></td>
    </tr>
    <tr>
      <th>write_to_file</th>
      <td><p>Logical indicating whether output should be written to a
.txt file (in current working directory). If <code>to_file = FALSE</code> the
function returns output R object. Defaults to <code>FALSE</code>.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A data frame with 1001 columns:</p><ol>
<li><p><code>gene</code> First column with the gene IDs (if found in input gff)</p></li>
<li><p><code>t1 to t1000</code> 1000 columns with signal at each meta position.</p></li>
</ol><p>Columns t1 to t1000 correspond to the following meta positions:</p><ol>
<li><p><code>t1 to t250</code> Upstream flank (1/2 of gene length)</p></li>
<li><p><code>t251 to t750</code> ORF</p></li>
<li><p><code>t751 to t1000</code> Downstream flank (1/2 of gene length)</p></li>
</ol>

    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='fu'>not_run</span>({
  <span class='fu'>signal_at_position</span>(<span class='no'>WT</span>, <span class='no'>Red1_summits</span>, <span class='st'>'name'</span>)

  <span class='fu'>signal_at_position</span>(<span class='no'>WT</span>, <span class='no'>centromeres</span>, <span class='st'>'seqnames'</span>,
                     <span class='kw'>up_ext</span> <span class='kw'>=</span> <span class='fl'>5000</span>, <span class='kw'>down_ext</span> <span class='kw'>=</span> <span class='fl'>5000</span>,
                     <span class='kw'>write_to_file</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>)
})</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Luis Vale Silva.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  </body>
</html>
