<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Signal at all ORFs genome-wide normalized to constant length (meta ORF) — signal_at_orf2 • hwglabr2</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://maxcdn.bootstrapcdn.com/bootswatch/3.3.7/yeti/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">

<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">


<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script>
<script src="../pkgdown.js"></script>
  
  
<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">hwglabr2</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../reference/index.html">Functions</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/hochwagenlab/hwglabr2">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

      <div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Signal at all ORFs genome-wide normalized to constant length (meta ORF)</h1>
    </div>

    
    <p>Pulls out the ChIP signal over all ORFs in the genome normalized to a
constant length. Collects the signal over each ORF plus both flanking regions
(1/2 the length of the ORF on each side) and scales them all to the same
value (1000). This means that for two example genes with lengths of 500 bp
and 2 kb, flanking regions of 250 bp and 1 kb will be collected up and
downstream, respectively. The whole region is then rescaled to a length of
1000, corresponding to a gene length of 500 plus 250 for each flanking
region.</p>
    

    <pre class="usage"><span class='fu'>signal_at_orf2</span>(<span class='no'>signal_data</span>, <span class='no'>gff</span>, <span class='kw'>window_size</span> <span class='kw'>=</span> <span class='fl'>1</span>, <span class='kw'>write_to_file</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='no'>file_name</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a> Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>signal_data</th>
      <td><p>Input signal track data as a <code>GRanges</code> object (see
<code>?"GRanges-class"</code> for more details). To load wiggle and bedGraph data
run <code><a href='import_wiggle.html'>import_wiggle</a></code> and <code><a href='import_bedGraph.html'>import_bedGraph</a></code>,
respectively. No default.</p></td>
    </tr>
    <tr>
      <th>gff</th>
      <td><p>Either a path to a gff file or loaded gff data as a <code>GRanges</code>
object. No default.</p></td>
    </tr>
    <tr>
      <th>window_size</th>
      <td><p>Integer specifying the width in bp of a window to average
signal in. Defaults to <code>1</code> (no averaging).</p></td>
    </tr>
    <tr>
      <th>write_to_file</th>
      <td><p>Logical indicating whether output should be written to a
.txt file. The file will be saved at location provided in <code>file_name</code>
argument or at the current working directory if only a file name is provided).
If <code>write_to_file = FALSE</code> the function returns output R object.
Defaults to <code>FALSE</code>.</p></td>
    </tr>
    <tr>
      <th>file_name</th>
      <td><p>Character string indicating file name when writing output to
a .txt file. Must be provided if <code>write_to_file = TRUE</code>. No default.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>An <code>EnrichedHeatmap</code> matrix with 1000 columns containing signal
at each meta position. Columns correspond to the following meta positions:</p><ol>
<li><p><code>t1 to t250</code> Upstream flank (1/2 of gene length)</p></li>
<li><p><code>t251 to t750</code> ORF</p></li>
<li><p><code>t751 to t1000</code> Downstream flank (1/2 of gene length)</p></li>
</ol>

    
    <h2 class="hasAnchor" id="warning"><a class="anchor" href="#warning"></a>Warning</h2>

    
    <p>Our ChIP-seq data always contains gaps with missing data. The
affected genes will contain "NA" values in the output. As a warning, the
number of affected genes is printed to the console.</p>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><span class='co'># NOT RUN {</span>
<span class='fu'>signal_at_orf2</span>(<span class='no'>WT</span>, <span class='kw'>gff</span> <span class='kw'>=</span> <span class='no'>gff</span>)

<span class='fu'>signal_at_orf2</span>(<span class='no'>WT</span>, <span class='kw'>gff</span> <span class='kw'>=</span> <span class='st'>"S288C_annotation_modified.gff"</span>,
               <span class='kw'>write_to_file</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>file_name</span> <span class='kw'>=</span> <span class='st'>"WT_S288C_metaORF.txt"</span>)
<span class='co'># }</span></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#warning">Warning</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by <a href='http://luisvalesilva.com'>Luis Vale Silva</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  </body>
</html>
